### PR TITLE
fix missed rbac for metrics-server

### LINKED ISSUE #118123: metrics-server's pod nanny broken after v0.6.3 upgrade
### What happened?

https://github.com/kubernetes/kubernetes/pull/117120 broke rbac permissions for pod nanny and it's now not able to scale up with number of nodes.

We see in kube-apiserver logs:
```
➜  latest grep nanny kube-apiserver.log
I0517 17:09:42.416260      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="395.912µs" userAgent="pod_nanny/1.8.14" audit-ID="c22e16af-03da-4ca2-b314-da1b018ed15d" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="101.716µs" resp=403
I0517 17:10:12.516158      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="567.486µs" userAgent="pod_nanny/1.8.14" audit-ID="e7b31723-4d60-4339-8769-911188ff95e7" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="187.707µs" resp=403
I0517 17:10:42.517920      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="554.668µs" userAgent="pod_nanny/1.8.14" audit-ID="5b4e4c63-bb9f-4293-bd2d-2e62409adc95" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="133.04µs" resp=403
I0517 17:11:12.519884      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="447.983µs" userAgent="pod_nanny/1.8.14" audit-ID="1154d19e-7b22-4fec-afa3-a355a53fe1da" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="74.523µs" resp=403
I0517 17:11:42.525213      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="415.112µs" userAgent="pod_nanny/1.8.14" audit-ID="bc3da8bf-2ae7-40af-a5a7-86e3123ee9a4" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="137.509µs" resp=403
I0517 17:12:12.528229      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="506.312µs" userAgent="pod_nanny/1.8.14" audit-ID="6259b896-33c0-4a7a-9bbd-7e50ca52c304" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="102.895µs" resp=403
I0517 17:12:42.530172      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="375.156µs" userAgent="pod_nanny/1.8.14" audit-ID="5fc6a91b-8a21-409c-bb5e-5b61f35d867d" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="107.313µs" resp=403
I0517 17:13:12.532099      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="305.642µs" userAgent="pod_nanny/1.8.14" audit-ID="649279ae-c9aa-4bd8-90f2-0fbd8e58c9f5" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="100.109µs" resp=403
I0517 17:13:42.542181      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="765.602µs" userAgent="pod_nanny/1.8.14" audit-ID="550d8190-cca6-42d7-81c7-0441b6912d3e" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="234.628µs" resp=403
I0517 17:14:12.543649      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="397.699µs" userAgent="pod_nanny/1.8.14" audit-ID="a75ba781-c795-43d2-9fd0-eaf589eafe43" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="112.211µs" resp=403
I0517 17:14:42.545270      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="376.4µs" userAgent="pod_nanny/1.8.14" audit-ID="c653face-6bcd-402e-a47c-a09a2fce1071" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="105.94µs" resp=403
(..)
I0517 18:50:13.694973      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="1.123175ms" userAgent="pod_nanny/1.8.14" audit-ID="8542e761-394d-46f8-90a0-ca330987eb63" srcIP="10.64.1.5:52092" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="120.336µs" resp=403
```

I tried to fix naively the problem by https://github.com/kubernetes/kubernetes/pull/118122 but it's incomplete as we still see:

```
I0519 08:27:00.038780      11 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="75.175445ms" userAgent="pod_nanny/1.8.14" audit-ID="65132840-b27b-4cff-acce-fe5781eb79fe" srcIP="10.64.1.4:56880" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="74.768907ms" resp=200
I0519 08:27:03.058724      11 httplog.go:132] "HTTP" verb="GET" URI="/api/v1/namespaces/kube-system/pods/metrics-server-v0.6.3-6d58b7896b-jkpvr" latency="2.218787ms" userAgent="pod_nanny/1.8.14" audit-ID="128e2ccb-4447-41e5-87e5-f2baeb2c163e" srcIP="10.64.1.4:56880" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="1.827883ms" resp=200
I0519 08:27:03.059997      11 httplog.go:132] "HTTP" verb="GET" URI="/apis/apps/v1/namespaces/kube-system/deployments/metrics-server-v0.6.3" latency="396.57µs" userAgent="pod_nanny/1.8.14" audit-ID="e539e7e9-3a6b-467c-8ca3-4b32e16d1582" srcIP="10.64.1.4:56880" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="110.62µs" resp=403
```

which means that we are missing more permissions (e.g. get deployments).

Sample test run where we see it: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1658880327533203456

At this stage I gave up.

/sig instrumentation
/cc @serathius @dgrisonnet 

### What did you expect to happen?

pod_nany has access to required data to scale up metrics-server

### How can we reproduce it (as minimally and precisely as possible)?

Create empty cluster, watch kube-apiserver logs for pod_nanny user agent

Also you can test e2e by calling '/test pull-kubernetes-e2e-gce-100-performance' and seeing if metrics-server ever resizes.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
master
</details>

### Cloud provider

<details>
gce
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by yangjunmyfm192085**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
fix https://github.com/kubernetes/kubernetes/issues/118123
#### Which issue(s) this PR fixes:

Fixes #118123

#### Does this PR introduce a user-facing change?

```release-note
 "NONE" 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
 "NONE" 
```

### CODE REVIEW

**Review by yangjunmyfm192085 — COMMENTED** [review]
/test pull-kubernetes-node-e2e-containerd

File: cluster/addons/metrics-server/resource-reader.yaml [hunk 1]
```diff
@@ -11,6 +11,20 @@ rules:
       - nodes/metrics
     verbs:
       - get
+  - apiGroups:
+      - "apps"
+    resources:
+      - deployments
+    verbs:
```
Comment (BenTheElder) [inline cluster/addons/metrics-server/resource-reader.yaml:18]: doesn't upstream only have get and patch?
and scoped to metrics-server only?
https://github.com/kubernetes-sigs/metrics-server/blob/1a23b5bd2e12ca55372c2cb6209c7eb1d94a2b4a/manifests/components/autoscale/rbac.yaml#L43

Comment (yangjunmyfm192085) [reply cluster/addons/metrics-server/resource-reader.yaml:20]: > doesn't upstream only have get and patch? and scoped to metrics-server only? https://github.com/kubernetes-sigs/metrics-server/blob/1a23b5bd2e12ca55372c2cb6209c7eb1d94a2b4a/manifests/components/autoscale/rbac.yaml#L43

Thanks for the review, I have updated

**Comment by mengjiao-liu** [comment]
/test pull-kubernetes-e2e-gce-100-performance

for seeing if metrics-server ever resizes.

**Comment by dgrisonnet** [comment]
Thank you @mengjiao-liu for running the test! :)

I looked at the kube-apiserver logs from the scale tests and I am still seeing 403 responses from nanny when getting the metrics-server deployment:
```
➜  e2e-118126-95a39-master git:(master) ✗ grep nanny kube-apiserver.log
I0529 07:09:03.318506      10 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="74.632354ms" userAgent="pod_nanny/1.8.14" audit-ID="ba24fd9d-d77a-4cd9-8678-16b44a6cb902" srcIP="10.64.1.5:39122" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="74.071486ms" resp=200
I0529 07:09:06.241526      10 httplog.go:132] "HTTP" verb="GET" URI="/api/v1/namespaces/kube-system/pods/metrics-server-v0.6.3-6d58b7896b-tk7xw" latency="2.607519ms" userAgent="pod_nanny/1.8.14" audit-ID="983cbe80-f2c8-440b-acde-74a703a30ae0" srcIP="10.64.1.5:39122" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="2.090453ms" resp=200
I0529 07:09:06.243150      10 httplog.go:132] "HTTP" verb="GET" URI="/apis/apps/v1/namespaces/kube-system/deployments/metrics-server-v0.6.3" latency="536.313µs" userAgent="pod_nanny/1.8.14" audit-ID="0dac5625-3fb9-42b6-8d59-1ed631c3542b" srcIP="10.64.1.5:39122" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="152.301µs" resp=403
I0529 07:09:36.410935      10 httplog.go:132] "HTTP" verb="GET" URI="/metrics" latency="76.87357ms" userAgent="pod_nanny/1.8.14" audit-ID="d1d00dee-46ae-4f01-ab6e-3ef9ad2cfe6b" srcIP="10.64.1.5:39122" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="76.408267ms" resp=200
```

I am not too sure what is missing since nanny should have the permissions to get the metrics-server deployment, I'll try to investigate further.

For anyone who wants to have a look at the logs, you can download them by running:
```
gsutil cp -r gs://sig-scalability-logs/pull-kubernetes-e2e-gce-100-performance/1663074369787662336 .
```

**Comment by dgrisonnet** [comment]
I fixed the issue in https://github.com/kubernetes/kubernetes/pull/119386.

/close

**Pull request closed without merging**